guppylang/std/_internal/compiler/array.py

```diff
@@ -180,44 +186,158 @@ def compile(self, args: list[Wire]) -> list[Wire]:
             return [self.build_classical_array(args)]
 
 
+ARRAY_GETITEM_CLASSICAL: Final[GlobalConstId] = GlobalConstId.fresh()
+ARRAY_GETITEM_LINEAR: Final[GlobalConstId] = GlobalConstId.fresh()
+ARRAY_SETITEM_CLASSICAL: Final[GlobalConstId] = GlobalConstId.fresh()
+ARRAY_SETITEM_LINEAR: Final[GlobalConstId] = GlobalConstId.fresh()
+
+
 class ArrayGetitemCompiler(ArrayCompiler):
     """Compiler for the `array.__getitem__` function."""
 
-    def build_classical_getitem(self, array: Wire, idx: Wire) -> CallReturnWires:
-        """Lowers a call to `array.__getitem__` for classical arrays."""
+    def _construct_getitem_ty(self, bound: ht.TypeBound) -> ht.PolyFuncType:
+        """Constructs a polymorphic function type for `__getitem__`"""
+        # a(T, N), int -> T, a(T, N)
+        # Array element type parameter
+        elem_ty_param = ht.TypeTypeParam(bound)
+        # Array length parameter
+        length_param = ht.BoundedNatParam()
+        return ht.PolyFuncType(
+            params=[elem_ty_param, length_param],
+            body=ht.FunctionType(
+                input=[
+                    array_type(
+                        ht.Option(ht.Variable(0, bound)),
+                        ht.VariableArg(1, length_param),
+                    ),
+                    INT_T,
+                ],
+                output=[
+                    ht.Variable(0, bound),
+                    array_type(
+                        ht.Option(ht.Variable(0, bound)),
+                        ht.VariableArg(1, length_param),
+                    ),
+                ],
+            ),
+        )
+
+    def _construct_classical_getitem(self, name: str) -> Wire:
+        """Constructs a generic function for `__getitem__` for classical arrays."""
+        func_ty = self._construct_getitem_ty(ht.TypeBound.Copyable)
+        parent_op = ops.FuncDefn(
+            name, func_ty.body.input, func_ty.params, func_ty.body.output
+        )
+        func = hf.Function.new_nested(parent_op, self.builder.hugr)
```

Comment:
This places the function into the current builder, but it should be on module level. This should fail if you try to use arrays in two different functions (maybe also add a test for that!)

Also no need to define the op by hand. You can try something like this:

```suggestion
        func = self.ctx.module.define_function(
           name, func_type.body.input, func_type.body.output, func_type.params
        )
```